server/src/main/java/org/elasticsearch/action/bulk/BulkPrimaryExecutionContext.java

```diff
@@ -0,0 +1,344 @@
+        currentItemState = ItemProcessingState.INITIAL;
+        currentIndex =  findNextNonAborted(currentIndex + 1);
+        retryCounter = 0;
+        requestToExecute = null;
+        executionResult = null;
+        assert assertInvariants(ItemProcessingState.INITIAL);
+    }
+
+    /** gets the current, untranslated item request */
+    public DocWriteRequest<?> getCurrent() {
+        return getCurrentItem().request();
+    }
+
+    public BulkShardRequest getBulkShardRequest() {
+        return request;
+    }
+
+    /** returns the result of the request that has been executed on the shard */
+    public BulkItemResponse getExecutionResult() {
+        assert assertInvariants(ItemProcessingState.EXECUTED);
+        return executionResult;
+    }
+
+    /** returns the number of times the current operation has been retried */
+    public int getRetryCounter() {
+        return retryCounter;
+    }
+
+    /** returns true if the current request has been executed on the primary */
+    public boolean isOperationExecuted() {
+        return currentItemState == ItemProcessingState.EXECUTED;
+    }
+
+    /** returns true if the request needs to wait for a mapping update to arrive from the master */
+    public boolean requiresWaitingForMappingUpdate() {
+        return currentItemState == ItemProcessingState.WAIT_FOR_MAPPING_UPDATE;
+    }
+
+    /** returns true if the current request should be retried without waiting for an external event */
+    public boolean requiresImmediateRetry() {
+        return currentItemState == ItemProcessingState.IMMEDIATE_RETRY;
+    }
+
+    /**
+     * returns true if the current request has been completed and it's result translated to a user
+     * facing response
+     */
+    public boolean isCompleted() {
+        return currentItemState == ItemProcessingState.COMPLETED;
+    }
+
+    /**
+     * returns true if the current request is in INITIAL state
+     */
+    public boolean isInitial() {
+        return currentItemState == ItemProcessingState.INITIAL;
+    }
+
+    /**
+     * returns true if {@link #advance()} has moved the current item beyond the
+     * end of the {@link BulkShardRequest#items()} array.
+     */
+    public boolean hasMoreOperationsToExecute() {
+        return currentIndex < request.items().length;
+    }
+
+
+    /** returns the name of the index the current request used */
+    public String getConcreteIndex() {
+        return getCurrentItem().index();
+    }
+
+    /** returns any primary response that was set by a previous primary */
+    public BulkItemResponse getPreviousPrimaryResponse() {
+        return getCurrentItem().getPrimaryResponse();
+    }
+
+    /** returns a translog location that is needed to be synced in order to persist all operations executed so far */
+    public Translog.Location getLocationToSync() {
+        assert hasMoreOperationsToExecute() == false;
+        assert assertInvariants(ItemProcessingState.INITIAL, ItemProcessingState.COMPLETED);
```

Comment:
I would have expected for this to only be INITIAL?

Reply:
if you have a bulk with all aboreted items you can overflow in advance and have initial here.

Reply:
My comment was that we should always end up in INITIAL, because we always move to INITIAL after completing an item

Reply:
GRR. Misread your comment. That is correct as far as I can tell. I pushed 5eeb932492fd4a147cfea05a0803a15082986959